community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/ast/rewriters/recordScopes.scala

```diff
@@ -19,15 +19,17 @@
  */
 package org.neo4j.cypher.internal.compiler.v3_1.ast.rewriters
 
-import org.neo4j.cypher.internal.frontend.v3_1.ast.PatternComprehension
-import org.neo4j.cypher.internal.frontend.v3_1.{Rewriter, SemanticState, bottomUp}
+import org.neo4j.cypher.internal.frontend.v3_1.ast.{MapProjection, PatternComprehension}
+import org.neo4j.cypher.internal.frontend.v3_1.{Rewriter, SemanticState, topDown}
 
 case class recordScopes(semanticState: SemanticState) extends Rewriter {
 
-  def apply(that: AnyRef): AnyRef = bottomUp(instance).apply(that)
+  def apply(that: AnyRef): AnyRef = topDown(instance).apply(that)
```

Comment:
Since `topDown()` creates an object, isn't it better to have that be done once, on the instance, instead of at each call to `apply()`? Ie:

```
def apply(that: AnyRef): AnyRef = instance.apply(that)

private val instance: Rewriter = topDown(Rewriter.lift { ...
```